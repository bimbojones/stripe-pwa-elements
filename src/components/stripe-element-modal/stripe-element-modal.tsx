import { Component, Host, h, Prop, Method, Element, Event, EventEmitter } from '@stencil/core';
import { checkPlatform } from '../../utils/utils';

@Component({
  tag: 'stripe-sheet',
  styleUrl: 'stripe-element-modal.scss',
  shadow: true,
})
export class StripeSheet {
  @Element() el: HTMLStripeSheetElement;

  /**
   * If true, the modal display close button
   */
  @Prop() showCloseButton = true;

  /**
   * Modal state.
   * If true, the modal will open
   */
  @Prop() open = false;

  /**
   * Toggle modal state
   */
  @Method()
  public async toggleModal() {
    this.open = !this.open;
    if (this.open === false) {
      this.close.emit();
    }
  }

  /**
   * Open the modal
   */
  @Method()
  public async openModal() {
    this.open = true;
  }

  /**
   * Close the modal
   */
  @Method()
  public async closeModal() {
    this.open = false;
    this.close.emit();
  }

  /**
   *
   */
  @Event() close: EventEmitter;

  componentDidLoad() {
    const platformName = checkPlatform()
    if (platformName) this.el.classList.add(platformName);
  }

  render() {
    const { open, showCloseButton } = this;

    return (
      <Host>
        <div class={`modal-row${open ? ' open' : ''}`} onClick={() => this.closeModal()}>
          <div class="modal-child" onClick={e => e.stopPropagation()}>
            {showCloseButton ? (
              <div class="modal-close-button-wrap">
                <ion-icon name="close" size="large" class="modal-close-button" onClick={() => this.closeModal()}></ion-icon>
              </div>
            ) : null}
            <slot></slot>
          </div>
        </div>
      </Host>
    );
  }
}
