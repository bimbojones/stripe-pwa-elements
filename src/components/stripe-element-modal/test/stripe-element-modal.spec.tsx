import { newSpecPage } from '@stencil/core/testing';
import { StripeSheet } from '../stripe-element-modal';

describe('stripe-sheet', () => {
  describe('Component Logic test', () => {
    describe('#openModal', () => {
      it('open props should change to true when the openModal method was called', async () => {
        const component = new StripeSheet();

        component.open = false;
        await component.openModal();
        expect(component.open).toEqual(true);
      });
    });
    describe('#closeModal', () => {
      it('open props should change to false when the closeModal method was called', async () => {
        const component = new StripeSheet();

        component.close = {
          emit: jest.fn(),
        };
        await component.closeModal();
        expect(component.open).toEqual(false);
      });
      it('When the closeModal method called, should called close event at once', async () => {
        const component = new StripeSheet();

        component.open = true;
        const mockEmitter = jest.fn();

        component.close = {
          emit: mockEmitter,
        };
        await component.closeModal();
        expect(mockEmitter).toBeCalledTimes(1);
      });
    });
    describe('#toggleModal', () => {
      describe('Props open=true', () => {
        let component = new StripeSheet();
        let mockEmitter = jest.fn();

        beforeEach(() => {
          component = new StripeSheet();
          mockEmitter = jest.fn();
          component.open = true;
          component.close = {
            emit: mockEmitter,
          };
        });
        it('should execute the close event', async () => {
          await component.toggleModal();
          expect(mockEmitter).toBeCalledTimes(1);
        });
        it('Should change the open props to be false', async () => {
          await component.toggleModal();
          expect(component.open).toEqual(false);
        });
      });
      describe('Props open=false', () => {
        let component = new StripeSheet();
        let mockEmitter = jest.fn();

        beforeEach(() => {
          component = new StripeSheet();
          mockEmitter = jest.fn();
          component.open = false;
          component.close = {
            emit: mockEmitter,
          };
        });
        it('should execute the close event', async () => {
          await component.toggleModal();
          expect(mockEmitter).toBeCalledTimes(0);
        });
        it('Should change the open props to be true', async () => {
          await component.toggleModal();
          expect(component.open).toEqual(true);
        });
      });
    });
  });
  describe('Rendering test', () => {
    it('renders', async () => {
      const page = await newSpecPage({
        components: [StripeSheet],
        html: `<stripe-sheet></stripe-sheet>`,
      });

      expect(page.root).toEqualHtml(`
      <stripe-sheet>
        <mock:shadow-root>
          <div class="modal-row">
            <div class="modal-child">
              <div class="modal-close-button-wrap">
                <ion-icon class="modal-close-button" name="close" size="large"></ion-icon>
              </div>
              <slot></slot>
            </div>
          </div>
        </mock:shadow-root>
      </stripe-sheet>
      `);
    });

    it("should match snapshot (open='true')", async () => {
      const page = await newSpecPage({
        components: [StripeSheet],
        html: `<stripe-sheet open="true"></stripe-sheet>`,
      });

      expect(page.root).toMatchSnapshot();
    });

    it("should match snapshot (showCloseButton='true')", async () => {
      const page = await newSpecPage({
        components: [StripeSheet],
        html: `<stripe-sheet show-close-button="true"></stripe-sheet>`,
      });

      expect(page.root).toMatchSnapshot();
    });
    it("should match snapshot (showCloseButton='false')", async () => {
      const page = await newSpecPage({
        components: [StripeSheet],
        html: `<stripe-sheet show-close-button="false"></stripe-sheet>`,
      });

      expect(page.root).toMatchSnapshot();
    });
  });
});
