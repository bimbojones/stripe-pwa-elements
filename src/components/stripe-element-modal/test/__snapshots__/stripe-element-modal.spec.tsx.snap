// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stripe-sheet Rendering test should match snapshot (open='true') 1`] = `
<stripe-sheet open="true">
  <template shadowrootmode="open">
    <div class="modal-row open">
      <div class="modal-child">
        <div class="modal-close-button-wrap">
          <ion-icon class="modal-close-button" name="close" size="large"></ion-icon>
        </div>
        <slot></slot>
      </div>
    </div>
  </template>
</stripe-sheet>
`;

exports[`stripe-sheet Rendering test should match snapshot (showCloseButton='false') 1`] = `
<stripe-sheet show-close-button="false">
  <template shadowrootmode="open">
    <div class="modal-row">
      <div class="modal-child">
        <slot></slot>
      </div>
    </div>
  </template>
</stripe-sheet>
`;

exports[`stripe-sheet Rendering test should match snapshot (showCloseButton='true') 1`] = `
<stripe-sheet show-close-button="true">
  <template shadowrootmode="open">
    <div class="modal-row">
      <div class="modal-child">
        <div class="modal-close-button-wrap">
          <ion-icon class="modal-close-button" name="close" size="large"></ion-icon>
        </div>
        <slot></slot>
      </div>
    </div>
  </template>
</stripe-sheet>
`;
